plugin/pkg/admission/podpreset/exists/admission.go

```diff
@@ -0,0 +1,155 @@
+)
+
+func init() {
+	admission.RegisterPlugin(PluginName, func(io.Reader) (admission.Interface, error) {
+		return NewExists()
+	})
+}
+
+// exists is an implementation of admission.Interface.
+// It enforces that if a binding is being defined with a PodPresetTemplate
+// that the cluster actually has support for it.
+type exists struct {
+	*admission.Handler
+	discovery         discovery.DiscoveryInterface
+	sync.Mutex        // protects podPresetsChecked and podPresetsExists
+	podPresetsChecked bool
+	podPresetsExists  bool
+}
+
+var _ = scadmission.WantsKubeClientSet(&exists{})
+
+func (e *exists) supportsPodPresets() bool {
+	e.Lock()
+	defer e.Unlock()
+	return e.podPresetsExists
+}
+
+// checkClusterForPodPresets uses discovery to find if the cluster has the
+// right version of settings that supports PodPresets. It will return true
+// if we have successfully queried the Discovery API and hence we know
+// for sure if the cluster supports PodPresets or not.
+// In that case, it also sets podPresetsChecked to true and podPresetsExists
+// to true if the cluster supports PodPresets.
+func (e *exists) checkClusterForPodPresets() bool {
+	synced := false
+	e.Lock()
+	synced = e.podPresetsChecked
+	e.Unlock()
+	if synced {
+		return true
+	}
+
+	resourceList, err := e.discovery.ServerResourcesForGroupVersion(SupportedPodPresetVersion)
+
+	if err != nil && !errors.IsNotFound(err) {
+		// We don't know if the cluster supports PodPresets or not
+		glog.V(4).Infof("ServerResourcesForGroupVersion failed: %s", err)
+		return false
+
+	}
+
+	e.Lock()
+	defer e.Unlock()
+	e.podPresetsChecked = true
+
+	if err != nil && errors.IsNotFound(err) {
+		// No such resource, which means the cluster does not support PodPresets
+		glog.V(4).Infof("No PodPreset support in the cluster, not allowing PodPresetTemplates in Bindings")
+		return true
+	}
+
+	for _, resource := range resourceList.APIResources {
+		if resource.Name == "podpresets" {
+			glog.V(4).Infof("PodPreset support found, allowing PodPresetTemplates in Bindings")
+			e.podPresetsExists = true
+		}
+	}
+	return true
+}
+
+func (e *exists) Admit(a admission.Attributes) error {
+	// We only care about bindings
+	if a.GetResource().Group != servicecatalog.GroupName || a.GetResource().GroupResource() != servicecatalog.Resource("bindings") {
+		return nil
+	}
+	binding, ok := a.GetObject().(*servicecatalog.Binding)
+	if !ok {
+		return errors.NewBadRequest("Resource was marked with kind Binding but was unable to be converted")
+	}
+	if binding.Spec.AlphaPodPresetTemplate == nil {
+		return nil
+	}
+
+	// we need to wait until we have successfully talked to the cluster and
+	// determined if the cluster supports PodPresets or not.
+	if !e.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("%s is not yet ready to handle request", PluginName))
+	}
+
+	// Request is for creating a binding with pod preset, check to see if the cluster supports it
+	if e.supportsPodPresets() {
+		return nil
```

Comment:
okay to admit because we have one, and we support doing something with them, right?